Gems/Atom/Asset/Shader/Code/Source/Editor/CommonFiles/Preprocessor.cpp

```diff
@@ -83,124 +76,125 @@ namespace AZ
-                return Fputs_StaticHinge(asString, od);
+                return 0;
             }
+            // chose the proper stream
+            auto& selectedStream = od == MCPP_OUT ? s_currentInstance->m_outStream : s_currentInstance->m_errStream;
+            auto tellBefore = selectedStream.tellp();
+            // append that message to it
+            selectedStream << s;
+            return aznumeric_cast<int>(selectedStream.tellp() - tellBefore);
+        }
 
-            // entry point from mcpp. hijacking its output
-            static int Fputs_StaticHinge(const char* s, MCPP_OUTDEST od)
+        int McppBinder::Fprintf_StaticHinge(MCPP_OUTDEST od, const char* format, ...)
+        {
+            if (!OkToLog(od))
             {
-                if (!OkToLog(od))
-                {
-                    return 0;
-                }
-                // chose the proper stream
-                auto& selectedStream = od == MCPP_OUT ? s_currentInstance->m_outStream : s_currentInstance->m_errStream;
-                auto tellBefore = selectedStream.tellp();
-                // append that message to it
-                selectedStream << s;
-                return aznumeric_cast<int>(selectedStream.tellp() - tellBefore);
+                return 0;
             }
-
-            // entry point from mcpp. hijacking its output
-            static int Fprintf_StaticHinge(MCPP_OUTDEST od, const char* format, ...)
+            // run the formatting on stack memory first, in case it's enough
+            char localBuffer[DefaultFprintfBufferSize];
+
+            va_list args;
+            
+            va_start(args, format);
+            int count = azvsnprintf(localBuffer, DefaultFprintfBufferSize, format, args);
+            va_end(args);
+            
+            char* result = localBuffer;
+
+            // @result will be bound to @biggerData in case @localBuffer is not big enough.
+            std::unique_ptr<char[]> biggerData;
+            // ">=" is the right comparison because in case count == bufferSize
+            // We will need an extra byte to accomodate the '\0' ending character.
+            if (count >= DefaultFprintfBufferSize)
             {
-                if (!OkToLog(od))
-                {
-                    return 0;
-                }
-                // run the formatting on stack memory first, in case it's enough
-                constexpr int bufferSize = 256;
-                char localBuffer[bufferSize];
-                va_list args;
+                // There wasn't enough space in the local store.
+                count++; // vsnprintf returns a size that doesn't include the null character.
+                biggerData.reset(new char[count]);
+                result = &biggerData[0];
+                
+                // Remark: for MacOS & Linux it is important to call va_start again before
+                // each call to azvsnprintf. Not required for Windows.
                 va_start(args, format);
```

Comment:
This is not PAL-ified. What will happen if this code is reached in Windows? You comments make it seem like it will crash.

Reply:
Not required in windows as in "optional". The code has been validated for Windows too.